engine/src/pausables.rs

```diff
@@ -0,0 +1,229 @@
+use aurora_engine_precompiles::secp256k1::ECRecover;
+use aurora_engine_precompiles::{prepaid_gas, Berlin};
+use aurora_engine_types::types::Address;
+use aurora_engine_types::BTreeSet;
+use bitflags::bitflags;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+bitflags! {
+    /// Wraps unsigned integer where each bit identifies a different precompile.
+    #[derive(BorshSerialize, BorshDeserialize, Default)]
+    pub struct PrecompileFlags: u32 {
+        const ECRECOVER           = 0b000000000000001;
+        const SHA256              = 0b000000000000010;
+        const RIPEMD160           = 0b000000000000100;
+        const IDENTITY            = 0b000000000001000;
+        const MODEXP              = 0b000000000010000;
+        const BN256_ADD           = 0b000000000100000;
+        const BN256_MUL           = 0b000000001000000;
+        const BN256_PAIR          = 0b000000010000000;
+        const BLAKE2F             = 0b000000100000000;
+        const RANDOM_SEED         = 0b000001000000000;
+        const CURRENT_ACCOUNT     = 0b000010000000000;
+        const PREDECESSOR_ACCOUNT = 0b000100000000000;
+        const EXIT_TO_ETHEREUM    = 0b001000000000000;
+        const EXIT_TO_NEAR        = 0b010000000000000;
+        const PREPAID_GAS         = 0b100000000000000;
+    }
+}
+
+impl PrecompileFlags {
+    /// Checks if the precompile belonging to the `address` is marked as paused.
+    pub fn is_paused_by_address(&self, address: &Address) -> bool {
+        let precompile_flag = if address == &ECRecover::ADDRESS {
+            PrecompileFlags::ECRECOVER
+        } else if address == &SHA256::ADDRESS {
+            PrecompileFlags::SHA256
+        } else if address == &RIPEMD160::ADDRESS {
+            PrecompileFlags::RIPEMD160
+        } else if address == &Identity::ADDRESS {
+            PrecompileFlags::IDENTITY
+        } else if address == &ModExp::<Berlin>::ADDRESS {
+            PrecompileFlags::MODEXP
+        } else if address == &Bn256Add::<Berlin>::ADDRESS {
+            PrecompileFlags::BN256_ADD
+        } else if address == &Bn256Mul::<Berlin>::ADDRESS {
+            PrecompileFlags::BN256_MUL
+        } else if address == &Bn256Pair::<Berlin>::ADDRESS {
+            PrecompileFlags::BN256_PAIR
+        } else if address == &Blake2F::ADDRESS {
+            PrecompileFlags::BLAKE2F
+        } else if address == &RandomSeed::ADDRESS {
+            PrecompileFlags::RANDOM_SEED
+        } else if address == &CurrentAccount::ADDRESS {
+            PrecompileFlags::CURRENT_ACCOUNT
+        } else if address == &predecessor_account::ADDRESS {
+            PrecompileFlags::PREDECESSOR_ACCOUNT
+        } else if address == &exit_to_ethereum::ADDRESS {
+            PrecompileFlags::EXIT_TO_ETHEREUM
+        } else if address == &exit_to_near::ADDRESS {
+            PrecompileFlags::EXIT_TO_NEAR
+        } else if address == &prepaid_gas::ADDRESS {
+            PrecompileFlags::PREPAID_GAS
```

Comment:
To be fair, most of these should never be paused. Should be only:

`EXIT_TO_ETHEREUM`
`EXIT_TO_NEAR`

(and upcoming cross contract call precompile)

There shouldn't ever be a reason why we would need to pause anything else but those two.